ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -453,14 +460,33 @@ chainSelectionForBlock
   -> InvalidBlockPunishment m
   -> m (Point blk)
 chainSelectionForBlock cdb@CDB{..} blockCache hdr punish = do
-    (invalid, succsOf, lookupBlockInfo, curChain, tipPoint, ledgerDB)
-      <- atomically $ (,,,,,)
-          <$> (forgetFingerprint <$> readTVar cdbInvalid)
-          <*> VolatileDB.filterByPredecessor  cdbVolatileDB
-          <*> VolatileDB.getBlockInfo         cdbVolatileDB
-          <*> Query.getCurrentChain           cdb
-          <*> Query.getTipPoint               cdb
-          <*> LgrDB.getCurrent                cdbLgrDB
+    (invalid, succsOf', lookupBlockInfo, lookupBlockInfo', curChain, tipPoint, ledgerDB, loeFrag)
+      <- atomically $ do
+          (invalid, succsOf, lookupBlockInfo, curChain, tipPoint, ledgerDB, loeFrag) <-
+                (,,,,,,)
+            <$> (forgetFingerprint <$> readTVar cdbInvalid)
+            <*> VolatileDB.filterByPredecessor  cdbVolatileDB
+            <*> VolatileDB.getBlockInfo         cdbVolatileDB
+            <*> Query.getCurrentChain           cdb
+            <*> Query.getTipPoint               cdb
+            <*> LgrDB.getCurrent                cdbLgrDB
+            <*> readTVar                        cdbLoEFrag
+
+          -- Let these two functions ignore invalid blocks
+          let lookupBlockInfo' = ignoreInvalid    cdb invalid lookupBlockInfo
+              succsOf'         = ignoreInvalidSuc cdb invalid succsOf
+
+              -- Applying the @loeDiff@ would switch from @curChain@ to @loeAnchor@
+              loeFrag' = case AF.intersect curChain loeFrag of
```

Comment:
```suggestion
              -- has the same tip as loeFrag and the same anchor as curChain
              loeFrag' = case AF.intersect curChain loeFrag of
```

Reply:
Do we need to do this? Is it harmless for the `loeFrag'` to have an anchor that is an ancestor of the immutable tip?

Maybe `Diff.diff` requires its two arguments to have the same anchor?

Reply:
The main reason for the `AF.intersect` is to make sure that `curChain` and `loeFrag` *do* intersect; but it indeed seems fine to use an `loeFrag` that reaches beyond the immutable tip; I think it just seemed to me like a nice property that was easy to establish at this place.

Reply:
Totally fair. I think it's worth a comment explaining that.